volume/fs.go

```diff
@@ -0,0 +1,90 @@
+package volume
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/puppetlabs/wash/activity"
+	"github.com/puppetlabs/wash/plugin"
+)
+
+// FS presents a view of the filesystem of an Execable resource.
+type FS struct {
+	plugin.EntryBase
+	executor plugin.Execable
+}
+
+// NewFS creates a new FS entry with the given name, using the supplied executor to satisfy volume
+// operations.
+func NewFS(name string, executor plugin.Execable) *FS {
+	fs := &FS{EntryBase: plugin.NewEntry(name), executor: executor}
+	// Caching handled in List.
+	fs.DisableCachingFor(plugin.ListOp)
+	return fs
+}
+
+// List will attempt to list the filesystem of an Execable resource (the executor). It will list
+// a directory tree based on supplied configuration (defaulting to `/var/log` if not specified).
+func (d *FS) List(ctx context.Context) ([]plugin.Entry, error) {
+	return List(ctx, d, "")
+}
+
+func consumeExec(ctx context.Context, result plugin.ExecResult) (*bytes.Buffer, error) {
+	var buf bytes.Buffer
+	for chunk := range result.OutputCh {
+		if chunk.Err != nil {
+			activity.Record(ctx, "Error on exec: %v", chunk.Err)
+			return nil, chunk.Err
```

Comment:
Would be nice to indicate that this is a streaming error. Matches what we do in https://github.com/puppetlabs/wash/blob/master/api/error.go#L37.

Also, I think the early return may cause some issues with the channel closing because it wouldn't be able to read any data after the error so the caller would block.

Reply:
It will. I guess it's probably better to close the channel and have the writer error.

Reply:
Actually, can't close it because it's receive-only. So I think the context cancellation has to take care of it.

Reply:
Should we keep reading until the end, then return the error?

Reply:
Yeah, I think reading until the end makes the most sense and is simple enough.